### ISSUE TITLE
[demo][hook-callback][Chrome Canary 63.0.3239.0] Map.get() is very slow

### ISSUE THREAD

**Issue by t2ym** [t2ym/thin-hook#155]
[demo][hook-callback][Chrome Canary 63.0.3239.0] `Map.get()` is very slow

More than 100 times slower than that on Chrome 61.

Observe until the release.

**Comment by t2ym** [comment]
## Suspected Root Cause (not verified yet)
- The slow Map object `_globalObjects` has their objects registered in the order of `Object.getOwnPropertyDescriptors()`, which may generate the index of the Map object very inefficient way.  Randomization of the registration order may solve the issue.

**Comment by t2ym** [comment]
## Observation (Chrome Canary 63 only)
- The costs of searching (`Map.get()`) differ in various keys.
  - `_globalObject.get(Object)` is very slow while `_globalObject.get(firebase)` is fast.

**Comment by t2ym** [comment]
## Chromium Source Code Analysis

- In `JSMap`, [`CollectionsBuiltinsAssembler::FindOrderedHashTableEntry`](https://github.com/v8/v8/blame/31cde16eeeccad72239c3036c8547a44a33f75d8/src/builtins/builtins-collections-gen.cc#L683) linearly walk the chain of the looked-up hash for the key, which can cause many iterations of searching (`Map.prototype.get`), depending on the key and the hash table.
  - The code has not been checked against Chrome 61, which has no such issue.
- This can cause slow lookup of the key and the value in `Map.prorotype.get` due to the inappropriately long chain of the hash table.

https://codereview.chromium.org/2958063002
https://github.com/v8/v8/commit/aba708a1464043c73e7267a2baef500c2c6c0cc9
https://github.com/v8/v8/commit/4e9e81454049b59b256a71a604a9c36a341f923e
https://github.com/v8/v8/blame/31cde16eeeccad72239c3036c8547a44a33f75d8/src/builtins/builtins-collections-gen.cc#L683

**Comment by t2ym** [comment]
## Potential Workaround
- Get `constructor.name` and compare with `_global[constructor.name]`
- Overheads for Chrome 61
- `Map` would be practically meaningless if this inefficiency should not be fixed in Turbofan engine for future release versions of Chrome browsers

```javascript
        name = ctor.name;
        if (_global[name] !== ctor) {
          name = _globalObjects.get(ctor);
        }
```

**Comment by t2ym** [comment]
## More practical approach: This is practically effective.
```javascript
  const isSlowMap = (() => {
    const __globalObjectsGet = (o) => { let name = o instanceof Object ? o.name : undefined; if (o === _global[name]) { return name; } else { return _globalObjects.get(o); } };
    let startMap = Date.now();
    for (let i = 0; i < 100000; i++) {
      _globalObjects.get(Object)
    }
    let endMap = Date.now();
    let startProperty = Date.now();
    for (let i = 0; i < 100000; i++) {
      __globalObjectsGet(Object)
    }
    let endProperty = Date.now();
    return (endMap - startMap) > (endProperty - startProperty);
  })();
  const _globalObjectsGet = function _globalObjectsGet(o) {
    switch (o) {
    case Object:
    case Function:
    case Array:
      return o.name;
    default:
      return _globalObjects.get(o);
    }
  }
```

**Comment by t2ym** [comment]
## Benchmarks

| Chrome version | __hook__acl | . (op/s) | = (op/s) | () (op/s) | f (op/s) |
|:-----------:|:-----:|-----:|-----:|-----:|-----:|
| 61.0.3163.100 | 0.0.159 | 1,824,817 | 2,110,595  | 954,380  | 5,824,111  |
| 63.0.3239.0   | 0.0.159 | 642,756  | 644,870  | 315,547  | 11,494,252  |
| 61.0.3163.100 | 0.0.160 | 2,236,636  | 2,283,626  | 1,034,768  | 6,172,839  |
| 63.0.3239.100 | 0.0.160 | 2,123,593  | 2,382,086  | 1,034,447  | 10,695,187  |

**Comment by t2ym** [comment]
The root cause is not fixed. So Chrome status should be tracked.

**Comment by t2ym** [comment]
The root cause is fixed in Chrome Canary 64.0.3240.0

**Comment by t2ym** [comment]
If the root cause is fixed in the Chrome 62 release (not beta), the changes for the issue will be removed.
